### PR TITLE
Fix invalid `setter` parse

### LINKED ISSUE #12074: Parser fail on `errorRecovery` mode
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

```console
node -p "require('@babel/parser').parse('const a = {set name(){}}', {errorRecovery: true})"
    if (method.kind === "set" && method.params[method.params.length - 1].type === "RestElement") {
                                                                         ^

TypeError: Cannot read property 'type' of undefined
    at Parser.checkGetterSetterParams (@babel\parser\lib\index.js:10792:74)
    at Parser.parseObjectMethod (@babel\parser\lib\index.js:10800:12)
    at Parser.parseObjPropValue (@babel\parser\lib\index.js:10841:23)
    at Parser.parsePropertyDefinition (@babel\parser\lib\index.js:10772:10)
    at Parser.parseObjectLike (@babel\parser\lib\index.js:10664:25)
    at Parser.parseExprAtom (@babel\parser\lib\index.js:10198:23)
    at Parser.parseExprSubscripts (@babel\parser\lib\index.js:9844:23)
    at Parser.parseUpdate (@babel\parser\lib\index.js:9824:21)
    at Parser.parseMaybeUnary (@babel\parser\lib\index.js:9813:17)
    at Parser.parseExprOps (@babel\parser\lib\index.js:9683:23)
```

With `errorRecovery: true` parser didn't throw syntax error

**Input Code**

```js
const a = {set name(){}}
```

**Expected behavior**

Should throw a SyntaxError or continue parse.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

N/A

**Environment**

```
- Babel version(s): 7.11.5
- Node/npm version: Node.js 12
- OS: Windows 10
- Monorepo: no
- How you are using Babel: parser api
```

**Possible Solution**

N/A

**Additional context**

Context: https://github.com/prettier/prettier/issues/9226

### PR DESCRIPTION
**Pull request by fisker**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12074
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | No
| License                  | MIT

I'm new to this codebase, not sure is this enough. If I'm missing anything, just let me know.

Found another similar code [here](https://github.com/babel/babel/blob/a3893129ec01e0b2c9eb4b75cd0199bd527605ee/packages/babel-parser/src/plugins/estree.js#L120), not sure if it has similar issue, and not sure how to test it.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12076"></a>

### CODE REVIEW

File: packages/babel-parser/test/fixtures/core/object/invalid-setter/output.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
+  "type": "File",
+  "start":0,"end":15,"loc":{"start":{"line":1,"column":0},"end":{"line":1,"column":15}},
+  "errors": [
+    "SyntaxError: setter must have exactly one formal parameter (1:3)"
```
Comment (existentialism) [inline packages/babel-parser/test/fixtures/core/object/invalid-setter/output.json:5]: Unrelated, but I think we can improve this error message:

```
A setter must have exactly one parameter.
```

Or Typescript's:

```
A 'set' accessor must have exactly one parameter.
```

**Review by existentialism — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
@fisker That similar code is used when producing an ESTree-compliant AST. You can test it enabling the `"estree"` parser plugin (example: `packages/babel-parser/test/fixtures/estree/options.json`)

**Comment by fisker** [comment]
@nicolo-ribaudo Thanks for this guidance, but I'm not near the laptop I used to make this PR, and I don't want setup environment on another computer, it took me a while to make it run.  

Can you help me test/fix it?

**Comment by existentialism** [comment]
@fisker I got you

**Comment by fisker** [comment]
Thank you!

**Comment by fisker** [comment]
I wonder if [this part](https://github.com/babel/babel/blob/84ea6e450113efac78fa0b6ad8c3278d06bd6d5d/packages/babel-parser/src/plugins/flow.js#L1184) can use the new logic?

**Comment by existentialism** [comment]
@fisker I'd say we can do that in a follow-up, as it requires more breaking things up (`FunctionTypeAnnotation` don't have the rest element inside `params`).

**Comment by fisker** [comment]
Okay, thank you very much.

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]

**Pull request merged**